File: javascript/pickle-bucket-the-journey-begins-chapter-1/solution.js
// https://www.codewars.com/kata/5ce5df8dcb83dc0020d2bd8f

const FLAVORS = 4;

function getAverages (pickleCounts) {
  if (!pickleCounts || !pickleCounts instanceof Array) {
    throw(`Expected an array as input. Instead got: ${pickleCounts}\n`
          + "This exception should only be thrown if test cases are added with"
          + "null or non-array edge cases. These are NOT required to be handled.");
  }
  while (pickleCounts.length % FLAVORS || pickleCounts.length == 0) {
    pickleCounts.push(0);
  }
  const groupCount = pickleCounts.length / FLAVORS;
  return pickleCounts.reduce(bucketize, []).map(p => p / groupCount);
}

function bucketize(bucket, pickleCount, index) {
  let i = index % FLAVORS;
  bucket[i] = pickleCount + (bucket[i] || 0);
  return bucket;
}

File: rust/pickle-bucket-the-journey-begins-chapter-1/solution.rs
// https://www.codewars.com/kata/5ce5df8dcb83dc0020d2bd8f

fn get_averages (pickle_counts: Vec<u32>) -> Vec<f64> {
  if pickle_counts.is_empty() {
      return vec![0.0, 0.0, 0.0, 0.0];
  }
  
  let mut pickle_counts = pickle_counts;
  
  while pickle_counts.len() % 4 != 0 {
      pickle_counts.push(0);
  }
  
  let group_count = (pickle_counts.len() as f64 / 4.0).ceil();
  
  pickle_counts.chunks(4)
    .fold(vec![0,0,0,0], |a, b| a.iter()
        .zip(b)
        .map(|(c, d)| c + d)
        .collect()
    )
    .iter()
    .map(|count| count.clone() as f64 / group_count) 
    .collect()
}

